### PR TITLE
[SPARK-28227][SQL] Support TRANSFORM with aggregation.

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
## What changes were proposed in this pull request?
For Spark SQL, it can't support sql like :
`SELECT TRANSFORM ( d2, max(d1) as maxd1, cast(sum(d3) as string))`
`USING 'cat' AS (a,b,c)`
`FROM script_trans`
`WHERE d1 <= 100`
`GROUP BY d2`
`HAVING maxd1 > 0`

But in hive, it can support this kind SQL.
This makes our SQL migration difficult and complex。
This PR is to support use Aggregation with TRANSFORM and make SQL migration from Hive to Sparkeasier.

## How was this patch tested?

Added unit test.

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
@cloud-fan @gatorsmile @HyukjinKwon @jerryshao @wangyum  
Hi all, could you help to review this and give some advise.

**Comment by AngersZhuuuu** [comment]
@dongjoon-hyun  
Could you help to review this for me and  can you help me to @ who work on this part

**Comment by wangyum** [comment]
Could we implement `ScriptTransformation` in sql/core first([SPARK-15694](https://issues.apache.org/jira/browse/SPARK-15694))?

**Comment by AngersZhuuuu** [comment]
> Could we implement `ScriptTransformation` in sql/core first([SPARK-15694](https://issues.apache.org/jira/browse/SPARK-15694))?

**Comment by HyukjinKwon** [comment]
Yea, I agree with @wangyum's. Can we? @AngersZhuuuu

**Comment by AngersZhuuuu** [comment]
> Yea, I agree with @wangyum's. Can we? @AngersZhuuuu

Nowadays  I am doing for support SparkThriftServer to enable proxy client user's authentication and make a PR for master.  
#25179  

After that, I will focus on this problem.

**Comment by HyukjinKwon** [comment]
@AngersZhuuuu, let's fix https://github.com/apache/spark/pull/25028#issuecomment-510929154 first. Feel free to open another PR.

**Comment by AngersZhuuuu** [comment]
> @AngersZhuuuu, let's fix [#25028 (comment)](https://github.com/apache/spark/pull/25028#issuecomment-510929154) first. Feel free to open another PR.

It's OK.

**Comment by alfozan** [comment]
@AngersZhuuuu one small regression:

test query:
`MAP k / 10 USING 'cat' AS (one) from (select 10 as k);
`

**Results:**
with master:
`1.0`

with this PR:

Exception:
`Error in query: cannot resolve '`(k / 10)`' given input columns: [(CAST(k AS DOUBLE) / CAST(10 AS DOUBLE))];;`

This is simplified test case from `HiveCompatibilitySuite`:
https://github.com/apache/spark/blob/master/sql/hive/src/test/resources/ql/src/test/queries/clientpositive/mapreduce1.q

**Comment by AngersZhuuuu** [comment]
> @AngersZhuuuu one small regression:
> 
> test query:
> `MAP k / 10 USING 'cat' AS (one) from (select 10 as k); `
> 
> **Results:**
> with master:
> `1.0`
> 
> with this PR:
> 
> Exception:
> `Error in query: cannot resolve '(k / 10)' given input columns: [(CAST(k AS DOUBLE) / CAST(10 AS DOUBLE))];;`
> 
> This is a simplified test case from `HiveCompatibilitySuite`:
> https://github.com/apache/spark/blob/master/sql/hive/src/test/resources/ql/src/test/queries/clientpositive/mapreduce1.q

Yeah, i found this problem too. In some complex column expressions,  we can't find a direct attribute name for a col, so I use sql pretty name. But some complex col's final name is not  the same as `toPrettySQL`.  Then cause this problem.

**Comment by alfozan** [comment]
Regarding the issue in https://github.com/apache/spark/pull/25028#issuecomment-547230458

Instead of 
```
    val namedExpressions = expressions.map {
      case e: NamedExpression => e
      case e: Expression => UnresolvedAlias(e)
    }
```

How about also using a manual aliasing function to create an alias for complex column expressions (i.e. BinaryExpression) instead of relying on `toPrettySQL`
 
Example: 
```    
   val aliasFunc = (position: Int, e: Expression) => s"gen_alias_${position}"

    val namedExpressions = expressions.zipWithIndex.map {
      case (e: NamedExpression, _) => e
      case (e: BinaryExpression, index) => UnresolvedAlias(e, Some(aliasFunc(index, _)))
      case (e: Expression, _) => UnresolvedAlias(e)
    }
```

Test query runs successfully.

**Comment by AngersZhuuuu** [comment]
> Regarding the issue in [#25028 (comment)](https://github.com/apache/spark/pull/25028#issuecomment-547230458)
> 
> Instead of
> 
> ```
>     val namedExpressions = expressions.map {
>       case e: NamedExpression => e
>       case e: Expression => UnresolvedAlias(e)
>     }
> ```
> 
> How about also using a manual aliasing function to create an alias for complex column expressions (i.e. BinaryExpression) instead of relying on `toPrettySQL`
> 
> Example:
> 
> ```
>    val aliasFunc = (position: Int, e: Expression) => s"gen_alias_${position}"
> 
>     val namedExpressions = expressions.zipWithIndex.map {
>       case (e: NamedExpression, _) => e
>       case (e: BinaryExpression, index) => UnresolvedAlias(e, Some(aliasFunc(index, _)))
>       case (e: Expression, _) => UnresolvedAlias(e)
>     }
> ```
> 
> Test query runs successfully.

Good suggestion, make subquery 's output and transform's input with same alias.  Can solve this problem. Nice method.

**Comment by alfozan** [comment]
Follow up - I think It's even better to always use a manual aliasing function and not just for a subset of expressions:

```    
   val aliasFunc = (position: Int, e: Expression) => s"gen_alias_${position}"

    val namedExpressions = expressions.zipWithIndex.map {
      case (e: NamedExpression, _) => e
      case (e: Expression, index) => UnresolvedAlias(e, Some(aliasFunc(index, _)))
    }
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 1]
```diff
@@ -506,12 +511,71 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
         AttributeReference("value", StringType)()), true)
     }
 
+    val namedExpressions = expressions.map {
+      case e: NamedExpression => e
+      case e: Expression => UnresolvedAlias(e)
+    }
```
Comment (alfozan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: re discussion in https://github.com/apache/spark/pull/25028#issuecomment-547230458 and https://github.com/apache/spark/pull/25028#issuecomment-547246714 and 

Alternative suggestion:
```
   val aliasFunc = (position: Int, e: Expression) => s"gen_alias_${position}"

    val namedExpressions = expressions.zipWithIndex.map {
      case (e: NamedExpression, _) => e
      case (e: Expression, index) => UnresolvedAlias(e, Some(aliasFunc(index, _)))
    }
```

Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala]: It is ok even to rename all expression with `aliasFunc`  in this place.
After current busy job I will restart this pr and follow your nice suggestion.

**Comment by alfozan** [comment]
Another issue:

test query:
```
FROM (select 1 as key, 100 as value) src
MAP src.*, src.key, CAST(src.key / 10 AS INT), CAST(src.key % 10 AS INT), src.value
USING 'cat' AS (k, v, tkey, ten, one, tvalue);
```

Results:
with master:
`1	100	1	0	1	100`

with this PR:
`1	100	1	100	1	0`

This is a simplified test case from HiveCompatibilitySuite:
https://github.com/apache/spark/blob/master/sql/hive/src/test/resources/ql/src/test/queries/clientpositive/mapreduce7.q

**Comment by alfozan** [comment]
simpler test case:

```
FROM (select 1 as key, 100 as value) src
MAP src.*, CAST(src.key % 10 AS INT), src.value
USING 'cat' AS (k, v, one, tvalue);
```

**Comment by AngersZhuuuu** [comment]
> simpler test case:
> 
> ```
> FROM (select 1 as key, 100 as value) src
> MAP src.*, CAST(src.key % 10 AS INT), src.value
> USING 'cat' AS (k, v, one, tvalue);
> ```

Star cause the problem . Trying solve this .

**Comment by AngersZhuuuu** [comment]
> simpler test case:
> 
> ```
> FROM (select 1 as key, 100 as value) src
> MAP src.*, CAST(src.key % 10 AS INT), src.value
> USING 'cat' AS (k, v, one, tvalue);
> ```

Found the reason, in current mode of trasform, in here
https://github.com/apache/spark/blob/095f7b05fd7ae8ce0d8a82f0c4bc26aa92853762/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala#L1098

It will expand star with t.child.output, after t.child was analyzed, it will have `src.key`, `src.value`, `gen_alias_2` `src.value`  output, then expand method will make all t.child's output match `src` as transform's input. Then this error happened. 

change to 

```
      // If the script transformation input contains Stars, expand it.
      case t: ScriptTransformation if containsStar(t.input) =>
        t.copy(
          input = t.child.output
        )
```

This change is reasonable since transform's input is it's child's output.

**Comment by alfozan** [comment]
Another test query:

```
SELECT TRANSFORM(key, abs(key))
USING 'cat'
FROM (SELECT DISTINCT key FROM src);
```

The physical plan looks different with this PR applied. in particular, HashAggregate gets another output column, while it should be just one:

`HashAggregate(keys=[key#18], functions=[], output=[key#18, **another_output**])
`

**Comment by AngersZhuuuu** [comment]
> Another test query:
> 
> ```
> SELECT TRANSFORM(key, abs(key))
> USING 'cat'
> FROM (SELECT DISTINCT key FROM src);
> ```
> 
> The physical plan looks different with this PR applied. in particular, HashAggregate gets another output column, while it should be just one:
> 
> `HashAggregate(keys=[key#18], functions=[], output=[key#18, **another_output**]) `

Since we can solve every thing with path a `Seq(UnresolvedStar(None))` to` ScriptTrasnform`

**Comment by alfozan** [comment]
Here's what I mean:

test query:

```
SELECT TRANSFORM(key, abs(key))
USING 'cat'
FROM (SELECT DISTINCT key FROM src);
```

The physical plan without this PR (on master):
```
ScriptTransformation [key#7, abs(key#7)], cat, [key#9, value#10], HiveScriptIOSchema(List(),List(),Some(org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe),Some(org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe),List((field.delim,	)),List((field.delim,	)),Some(org.apache.hadoop.hive.ql.exec.TextRecordReader),Some(org.apache.hadoop.hive.ql.exec.TextRecordWriter),true)
+- *(2) HashAggregate(keys=[1#16], functions=[], output=[key#7])
   +- Exchange hashpartitioning(1#16, 1), true, [id=#27]
      +- *(1) HashAggregate(keys=[1 AS 1#16], functions=[], output=[1#16])
         +- *(1) Scan OneRowRelation[]
```



The physical plan with this PR:
```
ScriptTransformation [key#0, abs(key)#9], cat, [key#2, value#3], HiveScriptIOSchema(List(),List(),Some(org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe),Some(org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe),List((field.delim,	)),List((field.delim,	)),Some(org.apache.hadoop.hive.ql.exec.TextRecordReader),Some(org.apache.hadoop.hive.ql.exec.TextRecordWriter),true)
+- *(2) HashAggregate(keys=[1#10], functions=[], output=[key#0, abs(key)#9])
   +- Exchange hashpartitioning(1#10, 1), true, [id=#21]
      +- *(1) HashAggregate(keys=[1 AS 1#10], functions=[], output=[1#10])
         +- *(1) Scan OneRowRelation[]
```


Difference:
`+- *(2) HashAggregate(keys=[1#16], functions=[], output=[key#7])
`vs
`+- *(2) HashAggregate(keys=[1#10], functions=[], output=[key#0, abs(key)#9])
`

**Comment by alfozan** [comment]
Another issue:

query:
`MAP k / 10 USING 'cat' AS (aa) from (select 10 as k);
`

Error in query: cannot resolve '`(k / 10)`' given input columns: [(CAST(k AS DOUBLE) / CAST(10 AS DOUBLE))];;

**Comment by AngersZhuuuu** [comment]
> Another issue:
> 
> query:
> `MAP k / 10 USING 'cat' AS (aa) from (select 10 as k); `
> 
> Error in query: cannot resolve '`(k / 10)`' given input columns: [(CAST(k AS DOUBLE) / CAST(10 AS DOUBLE))];;

This can be solved by https://github.com/apache/spark/pull/25028#issuecomment-548675869

**Comment by alfozan** [comment]
> > Another issue:
> > query:
> > `MAP k / 10 USING 'cat' AS (aa) from (select 10 as k); `
> > Error in query: cannot resolve '`(k / 10)`' given input columns: [(CAST(k AS DOUBLE) / CAST(10 AS DOUBLE))];;
> 
> This can be solved by [#25028 (comment)](https://github.com/apache/spark/pull/25028#issuecomment-548675869)

Yes we already discussed a solution in https://github.com/apache/spark/pull/25028#discussion_r339890074
which I can confirm it works.

Currently the open issue is https://github.com/apache/spark/pull/25028#issuecomment-548691723

**Comment by AngersZhuuuu** [comment]
> output=[key#7]

This is right since we current tread transform's child logicalplan as an entire one.

**Comment by AngersZhuuuu** [comment]
@alfozan 
You can see my branch's newest change. and check the update

**Comment by AngersZhuuuu** [comment]
> Difference:
> `+- *(2) HashAggregate(keys=[1#16], functions=[], output=[key#7]) `vs
> `+- *(2) HashAggregate(keys=[1#10], functions=[], output=[key#0, abs(key)#9])`

Seems after my pr. the output is right.

**Comment by alfozan** [comment]
> @alfozan
> You can see my branch's newest change. and check the update

Could you share a link to the new branch/PR?

**Comment by AngersZhuuuu** [comment]
> > @alfozan
> > You can see my branch's newest change. and check the update
> 
> Could you share a link to the new branch/PR?

https://github.com/AngersZhuuuu/spark/commit/f18a6f99043ec26ac19a8337f41194b6f972090c

**Comment by alfozan** [comment]
Looks good! Thank you.

**Comment by AngersZhuuuu** [comment]
> Looks good! Thank you.

Also thank you for so many error case to make this pr better.

**Comment by AngersZhuuuu** [comment]
@alfozan 
Seems you are working on https://issues.apache.org/jira/browse/SPARK-15694
But I am confused that current ScripTransform execution process has been implemented, are you clear what we need to do?

**Comment by alfozan** [comment]
> @alfozan
> Seems you are working on https://issues.apache.org/jira/browse/SPARK-15694
> But I am confused that current ScripTransform execution process has been implemented, are you clear what we need to do?

yes - we presented the work in https://databricks.com/session_eu19/powering-custom-apps-at-facebook-using-spark-script-transformation and I'll send the PRs later this month.

**Comment by AngersZhuuuu** [comment]
> yes - we presented the work in https://databricks.com/session_eu19/powering-custom-apps-at-facebook-using-spark-script-transformation and I'll send the PRs later this month.

please ping me when you raise a pr to learn your great job.

**Comment by AngersZhuuuu** [comment]
@HyukjinKwon  
Can i reopen this since @alfozan is working or transform script.

**Comment by HyukjinKwon** [comment]
@AngersZhuuuu, I didn't read all fully but can we make https://issues.apache.org/jira/browse/SPARK-15694 done first as @wangyum's advice?

**Comment by AngersZhuuuu** [comment]
> @AngersZhuuuu, I didn't read all fully but can we make https://issues.apache.org/jira/browse/SPARK-15694 done first as @wangyum's advice?

Seems @alfozan is doing that part?

**Comment by HyukjinKwon** [comment]
Shall we wait for that first? I would like to see how it's implemented first, and possibly we should match the impl.

**Comment by alfozan** [comment]
> > @AngersZhuuuu, I didn't read all fully but can we make https://issues.apache.org/jira/browse/SPARK-15694 done first as @wangyum's advice?
> 
> Seems @alfozan is doing that part?

Yes I'm

**Pull request closed without merging**